pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,564 @@
+	*storageos
+}
+
+var _ volume.Unmounter = &storageosUnmounter{}
+
+func (b *storageosUnmounter) GetPath() string {
+	return getPath(b.podUID, b.namespace, b.volName, b.pvName, b.plugin.host)
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (b *storageosUnmounter) TearDown() error {
+
+	if len(b.namespace) == 0 || len(b.volName) == 0 {
+		return nil
+	}
+	// Unmount from pod
+	mountPath := b.GetPath()
+
+	err := b.TearDownAt(mountPath)
+	if err != nil {
+		glog.Errorf("storageos: Unmount from pod failed: %v", err)
+		return err
+	}
+
+	// Find device name from global mount
+	globalPDPath := makeGlobalPDName(b.plugin.host, b.namespace, b.volName)
+	devicePath, _, err := mount.GetDeviceNameFromMount(b.mounter, globalPDPath)
+	if err != nil {
+		glog.Errorf("Detach failed when getting device from global mount: %v", err)
+		return err
+	}
+
+	// Unmount from plugin's disk global mount dir.
+	err = b.TearDownAt(globalPDPath)
+	if err != nil {
+		glog.Errorf("Detach failed during unmount: %v", err)
+		return err
+	}
+
+	// Detach loop device
+	err = b.manager.DetachVolume(b, devicePath)
+	if err != nil {
+		glog.Errorf("Detach device %s failed for volume %s: %v", devicePath, b.pvName, err)
+		return err
+	}
+
+	glog.V(4).Infof("Successfully unmounted StorageOS volume %s and detached devices", b.pvName)
+
+	return nil
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (b *storageosUnmounter) TearDownAt(dir string) error {
+	if err := util.UnmountPath(dir, b.mounter); err != nil {
+		glog.V(4).Infof("Unmounted StorageOS volume %s failed with: %v", b.pvName, err)
+	}
+	if err := b.manager.UnmountVolume(b); err != nil {
+		glog.V(4).Infof("Mount reference for volume %s could not be removed from StorageOS: %v", b.pvName, err)
+	}
+	return nil
+}
+
+type storageosDeleter struct {
+	*storageos
+}
+
+var _ volume.Deleter = &storageosDeleter{}
+
+func (d *storageosDeleter) GetPath() string {
+	return getPath(d.podUID, d.namespace, d.volName, d.pvName, d.plugin.host)
+}
+
+func (d *storageosDeleter) Delete() error {
+	return d.manager.DeleteVolume(d)
+}
+
+type storageosProvisioner struct {
+	*storageos
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &storageosProvisioner{}
+
+func (v *storageosProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	// Set parameters from StorageClass
+	if len(v.options.Parameters) > 0 {
+		if description, ok := v.options.Parameters["description"]; ok {
+			v.description = description
+		}
+		if pool, ok := v.options.Parameters["pool"]; ok {
+			v.pool = pool
+		}
+		if fsType, ok := v.options.Parameters["fstype"]; ok {
+			v.fsType = fsType
+		}
+	}
+
+	// Set from PVC
+	v.volName = v.options.PVName
+	v.namespace = v.options.PVC.Namespace
+	v.labels = make(map[string]string)
+	for k, val := range v.options.PVC.Labels {
```

Comment:
We copy the labels from the PVC to the requested volume.  StorageOS uses the labels from k8s to set policy.  Fox example, if the PVC has environment=production set, a policy rule within StorageOS might set the number of replicas to 2.